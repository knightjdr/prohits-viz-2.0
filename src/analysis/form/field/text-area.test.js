import React from 'react';
import { mount } from 'enzyme';

import InfoModal from './info-modal';
import TextArea from './text-area';
import TestForm from './__mocks__/form-wrapper';
import UndefinedIfNotSet from '../../../helpers/undefined-if-not-set';

jest.mock('./info-modal');
jest.mock('../../../helpers/undefined-if-not-set');
UndefinedIfNotSet.mockImplementation(value => (value));

const inputOnChange = jest.fn();
const onChange = jest.fn();

describe('TextArea', () => {
  test('Renders with no value, help message and type = text', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
          rows={5}
        />
      </TestForm>,
    );
    expect(wrapper).toMatchSnapshot();
    const input = wrapper.find('textarea');
    expect(input.props().defaultValue).toBeUndefined();
    expect(input.props().rows).toBe(5);
    expect(wrapper.find('svg.customfield__help').length).toBe(1);
  });

  test('Renders with without help message', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
        />
      </TestForm>,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('svg.customfield__help').length).toBe(0);
  });

  test('Modal called on button click', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          label="Label"
          input={{}}
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
        />
      </TestForm>,
    );
    jest.clearAllMocks();
    const button = wrapper.find('.customfield__help').first();
    button.simulate('click');
    expect(InfoModal).toHaveBeenCalledTimes(1);
    expect(InfoModal).toHaveBeenCalledWith('Label', 'help');
  });

  test('Modal called on button click without title text', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
        />
      </TestForm>,
    );
    jest.clearAllMocks();
    const button = wrapper.find('.customfield__help').first();
    button.simulate('click');
    expect(InfoModal).toHaveBeenCalledTimes(1);
    expect(InfoModal).toHaveBeenCalledWith('Help', 'help');
  });

  test('On change called on blur when input changes', () => {
    const formTextArea = {
      onChange: inputOnChange,
      value: undefined,
    };
    const wrapper = mount(
      <TestForm
        input={formTextArea}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
          type="number"
        />
      </TestForm>,
    );
    jest.clearAllMocks();
    const input = wrapper.find('textarea');
    input.props().onMouseLeave(10);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(10, formTextArea);
  });

  test('On change not called on blur when input is the same', () => {
    const formTextArea = {
      onChange: inputOnChange,
      value: undefined,
    };
    const wrapper = mount(
      <TestForm
        input={formTextArea}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
          type="number"
        />
      </TestForm>,
    );
    jest.clearAllMocks();
    const input = wrapper.find('textarea');
    input.props().onMouseLeave(undefined);
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  test('Can be set via input.value', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
          type="number"
        />
      </TestForm>,
    );
    wrapper.setProps({
      input: {
        onChange: inputOnChange,
        value: 1,
      },
    });
    const input = wrapper.find('textarea');
    expect(input.props().defaultValue).toBe(1);
  });

  test('Submit called on button click', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
        onSubmit={onSubmitSpy}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
          type="number"
        />
      </TestForm>,
    );
    const button = wrapper.find('button');
    button.simulate('submit');
    expect(onSubmitSpy).toHaveBeenCalledTimes(1);
  });

  test('Submit error adds prop visualization queue', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: 'Error message', touched: true, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{}}
          type="number"
        />
      </TestForm>,
    );
    expect(wrapper).toMatchSnapshot();
    const formItem = wrapper.find('FormItem');
    expect(formItem.props().help).toBe('Error message');
    expect(formItem.props().validateStatus).toBe('error');
  });

  test('Can add custom style', () => {
    const wrapper = mount(
      <TestForm
        input={{
          onChange: inputOnChange,
          value: undefined,
        }}
        meta={{ error: '', touched: false, warning: '' }}
      >
        <TextArea
          helpMessage="help"
          input={{}}
          label="Label"
          meta={{}}
          name="TestTextArea"
          onChange={onChange}
          placeHolder="TextArea..."
          style={{ backgroundColor: '#000' }}
          type="number"
        />
      </TestForm>,
    );
    const inputStyle = wrapper.find('textarea').first().props().style;
    expect(inputStyle).toHaveProperty('backgroundColor', '#000');
  });
});
