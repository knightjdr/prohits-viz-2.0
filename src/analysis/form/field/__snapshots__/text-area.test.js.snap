// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Renders with no value, help message and type = text 1`] = `
<TestForm
  input={
    Object {
      "onChange": [Function],
      "value": undefined,
    }
  }
  meta={
    Object {
      "error": "",
      "touched": false,
      "warning": "",
    }
  }
  onSubmit={[Function]}
>
  <Form
    hideRequiredMark={false}
    layout="horizontal"
    onSubmit={[Function]}
    prefixCls="ant-form"
  >
    <form
      className="ant-form ant-form-horizontal"
      onSubmit={[Function]}
    >
      <CustomInput
        helpMessage="help"
        input={
          Object {
            "onChange": [Function],
            "value": undefined,
          }
        }
        label="Label"
        meta={
          Object {
            "error": "",
            "touched": false,
            "warning": "",
          }
        }
        name="TestTextArea"
        onChange={[Function]}
        placeHolder="TextArea..."
        rows={5}
        style={Object {}}
      >
        <div
          className="CustomField-container"
        >
          <FormItem
            className="CustomField-formItem"
            colon={true}
            hasFeedback={false}
            help=""
            label="Label"
            prefixCls="ant-form"
            validateStatus=""
          >
            <Row
              className="ant-form-item CustomField-formItem"
              gutter={0}
            >
              <div
                className="ant-row ant-form-item CustomField-formItem"
              >
                <Col
                  className="ant-form-item-label"
                  key=".$label"
                >
                  <div
                    className="ant-form-item-label"
                  >
                    <label
                      className=""
                      htmlFor={null}
                      onClick={[Function]}
                      title="Label"
                    >
                      Label
                    </label>
                  </div>
                </Col>
                <Col
                  className="ant-form-item-control-wrapper"
                  key=".$wrapper"
                >
                  <div
                    className="ant-form-item-control-wrapper"
                  >
                    <div
                      className="ant-form-item-control"
                    >
                      <span
                        className="ant-form-item-children"
                      >
                        <TextArea
                          onMouseLeave={[Function]}
                          placeholder="TextArea..."
                          prefixCls="ant-input"
                          rows={5}
                          style={Object {}}
                        >
                          <textarea
                            className="ant-input"
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            onMouseLeave={[Function]}
                            placeholder="TextArea..."
                            rows={5}
                            style={Object {}}
                          />
                        </TextArea>
                      </span>
                      <Animate
                        animation={Object {}}
                        component=""
                        componentProps={Object {}}
                        key="help"
                        onAppear={[Function]}
                        onEnd={[Function]}
                        onEnter={[Function]}
                        onLeave={[Function]}
                        transitionAppear={true}
                        transitionEnter={true}
                        transitionLeave={true}
                        transitionName="show-help"
                      />
                    </div>
                  </div>
                </Col>
              </div>
            </Row>
          </FormItem>
          <FontAwesomeIcon$1
            border={false}
            className="CustomField-help"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f2fd",
                  "M400 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 430c-25.365 0-46-20.636-46-46 0-25.365 20.635-46 46-46s46 20.635 46 46c0 25.364-20.635 46-46 46zm40-131.333V300c0 6.627-5.373 12-12 12h-56c-6.627 0-12-5.373-12-12v-4c0-41.059 31.128-57.472 54.652-70.66 20.171-11.309 32.534-19 32.534-33.976 0-19.81-25.269-32.958-45.698-32.958-27.19 0-39.438 13.139-57.303 35.797-4.045 5.13-11.46 6.069-16.665 2.122l-34.699-26.31c-5.068-3.843-6.251-10.972-2.715-16.258C141.4 112.957 176.158 90 230.655 90c56.366 0 116.531 43.998 116.531 102 0 77.02-83.186 78.205-83.186 106.667z",
                ],
                "iconName": "question-square",
                "prefix": "fas",
              }
            }
            listItem={false}
            mask={null}
            name=""
            onClick={[Function]}
            pull={null}
            pulse={false}
            rotation={null}
            size="2x"
            spin={false}
            symbol={false}
            transform={null}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-question-square fa-w-14 fa-2x CustomField-help"
              data-icon="question-square"
              data-prefix="fas"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 430c-25.365 0-46-20.636-46-46 0-25.365 20.635-46 46-46s46 20.635 46 46c0 25.364-20.635 46-46 46zm40-131.333V300c0 6.627-5.373 12-12 12h-56c-6.627 0-12-5.373-12-12v-4c0-41.059 31.128-57.472 54.652-70.66 20.171-11.309 32.534-19 32.534-33.976 0-19.81-25.269-32.958-45.698-32.958-27.19 0-39.438 13.139-57.303 35.797-4.045 5.13-11.46 6.069-16.665 2.122l-34.699-26.31c-5.068-3.843-6.251-10.972-2.715-16.258C141.4 112.957 176.158 90 230.655 90c56.366 0 116.531 43.998 116.531 102 0 77.02-83.186 78.205-83.186 106.667z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </FontAwesomeIcon$1>
        </div>
      </CustomInput>
      <button
        type="submit"
      />
    </form>
  </Form>
</TestForm>
`;

exports[`TextArea Renders with without help message 1`] = `
<TestForm
  input={
    Object {
      "onChange": [Function],
      "value": undefined,
    }
  }
  meta={
    Object {
      "error": "",
      "touched": false,
      "warning": "",
    }
  }
  onSubmit={[Function]}
>
  <Form
    hideRequiredMark={false}
    layout="horizontal"
    onSubmit={[Function]}
    prefixCls="ant-form"
  >
    <form
      className="ant-form ant-form-horizontal"
      onSubmit={[Function]}
    >
      <CustomInput
        helpMessage={null}
        input={
          Object {
            "onChange": [Function],
            "value": undefined,
          }
        }
        label="Label"
        meta={
          Object {
            "error": "",
            "touched": false,
            "warning": "",
          }
        }
        name="TestTextArea"
        onChange={[Function]}
        placeHolder="TextArea..."
        rows={5}
        style={Object {}}
      >
        <div
          className="CustomField-container"
        >
          <FormItem
            className="CustomField-formItem"
            colon={true}
            hasFeedback={false}
            help=""
            label="Label"
            prefixCls="ant-form"
            validateStatus=""
          >
            <Row
              className="ant-form-item CustomField-formItem"
              gutter={0}
            >
              <div
                className="ant-row ant-form-item CustomField-formItem"
              >
                <Col
                  className="ant-form-item-label"
                  key=".$label"
                >
                  <div
                    className="ant-form-item-label"
                  >
                    <label
                      className=""
                      htmlFor={null}
                      onClick={[Function]}
                      title="Label"
                    >
                      Label
                    </label>
                  </div>
                </Col>
                <Col
                  className="ant-form-item-control-wrapper"
                  key=".$wrapper"
                >
                  <div
                    className="ant-form-item-control-wrapper"
                  >
                    <div
                      className="ant-form-item-control"
                    >
                      <span
                        className="ant-form-item-children"
                      >
                        <TextArea
                          onMouseLeave={[Function]}
                          placeholder="TextArea..."
                          prefixCls="ant-input"
                          rows={5}
                          style={Object {}}
                        >
                          <textarea
                            className="ant-input"
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            onMouseLeave={[Function]}
                            placeholder="TextArea..."
                            rows={5}
                            style={Object {}}
                          />
                        </TextArea>
                      </span>
                      <Animate
                        animation={Object {}}
                        component=""
                        componentProps={Object {}}
                        key="help"
                        onAppear={[Function]}
                        onEnd={[Function]}
                        onEnter={[Function]}
                        onLeave={[Function]}
                        transitionAppear={true}
                        transitionEnter={true}
                        transitionLeave={true}
                        transitionName="show-help"
                      />
                    </div>
                  </div>
                </Col>
              </div>
            </Row>
          </FormItem>
        </div>
      </CustomInput>
      <button
        type="submit"
      />
    </form>
  </Form>
</TestForm>
`;

exports[`TextArea Submit error adds prop visualization queue 1`] = `
<TestForm
  input={
    Object {
      "onChange": [Function],
      "value": undefined,
    }
  }
  meta={
    Object {
      "error": "Error message",
      "touched": true,
      "warning": "",
    }
  }
  onSubmit={[Function]}
>
  <Form
    hideRequiredMark={false}
    layout="horizontal"
    onSubmit={[Function]}
    prefixCls="ant-form"
  >
    <form
      className="ant-form ant-form-horizontal"
      onSubmit={[Function]}
    >
      <CustomInput
        helpMessage="help"
        input={
          Object {
            "onChange": [Function],
            "value": undefined,
          }
        }
        label="Label"
        meta={
          Object {
            "error": "Error message",
            "touched": true,
            "warning": "",
          }
        }
        name="TestTextArea"
        onChange={[Function]}
        placeHolder="TextArea..."
        rows={5}
        style={Object {}}
        type="number"
      >
        <div
          className="CustomField-container"
        >
          <FormItem
            className="CustomField-formItem"
            colon={true}
            hasFeedback={false}
            help="Error message"
            label="Label"
            prefixCls="ant-form"
            validateStatus="error"
          >
            <Row
              className="ant-form-item ant-form-item-with-help CustomField-formItem"
              gutter={0}
            >
              <div
                className="ant-row ant-form-item ant-form-item-with-help CustomField-formItem"
              >
                <Col
                  className="ant-form-item-label"
                  key=".$label"
                >
                  <div
                    className="ant-form-item-label"
                  >
                    <label
                      className=""
                      htmlFor={null}
                      onClick={[Function]}
                      title="Label"
                    >
                      Label
                    </label>
                  </div>
                </Col>
                <Col
                  className="ant-form-item-control-wrapper"
                  key=".$wrapper"
                >
                  <div
                    className="ant-form-item-control-wrapper"
                  >
                    <div
                      className="ant-form-item-control has-error"
                    >
                      <span
                        className="ant-form-item-children"
                      >
                        <TextArea
                          onMouseLeave={[Function]}
                          placeholder="TextArea..."
                          prefixCls="ant-input"
                          rows={5}
                          style={Object {}}
                        >
                          <textarea
                            className="ant-input"
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            onMouseLeave={[Function]}
                            placeholder="TextArea..."
                            rows={5}
                            style={Object {}}
                          />
                        </TextArea>
                      </span>
                      <Animate
                        animation={Object {}}
                        component=""
                        componentProps={Object {}}
                        key="help"
                        onAppear={[Function]}
                        onEnd={[Function]}
                        onEnter={[Function]}
                        onLeave={[Function]}
                        transitionAppear={true}
                        transitionEnter={true}
                        transitionLeave={true}
                        transitionName="show-help"
                      >
                        <AnimateChild
                          animation={Object {}}
                          key="help"
                          transitionAppear={true}
                          transitionEnter={true}
                          transitionLeave={true}
                          transitionName="show-help"
                        >
                          <div
                            className="ant-form-explain"
                            key="help"
                          >
                            Error message
                          </div>
                        </AnimateChild>
                      </Animate>
                    </div>
                  </div>
                </Col>
              </div>
            </Row>
          </FormItem>
          <FontAwesomeIcon$1
            border={false}
            className="CustomField-help"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f2fd",
                  "M400 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 430c-25.365 0-46-20.636-46-46 0-25.365 20.635-46 46-46s46 20.635 46 46c0 25.364-20.635 46-46 46zm40-131.333V300c0 6.627-5.373 12-12 12h-56c-6.627 0-12-5.373-12-12v-4c0-41.059 31.128-57.472 54.652-70.66 20.171-11.309 32.534-19 32.534-33.976 0-19.81-25.269-32.958-45.698-32.958-27.19 0-39.438 13.139-57.303 35.797-4.045 5.13-11.46 6.069-16.665 2.122l-34.699-26.31c-5.068-3.843-6.251-10.972-2.715-16.258C141.4 112.957 176.158 90 230.655 90c56.366 0 116.531 43.998 116.531 102 0 77.02-83.186 78.205-83.186 106.667z",
                ],
                "iconName": "question-square",
                "prefix": "fas",
              }
            }
            listItem={false}
            mask={null}
            name=""
            onClick={[Function]}
            pull={null}
            pulse={false}
            rotation={null}
            size="2x"
            spin={false}
            symbol={false}
            transform={null}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-question-square fa-w-14 fa-2x CustomField-help"
              data-icon="question-square"
              data-prefix="fas"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 430c-25.365 0-46-20.636-46-46 0-25.365 20.635-46 46-46s46 20.635 46 46c0 25.364-20.635 46-46 46zm40-131.333V300c0 6.627-5.373 12-12 12h-56c-6.627 0-12-5.373-12-12v-4c0-41.059 31.128-57.472 54.652-70.66 20.171-11.309 32.534-19 32.534-33.976 0-19.81-25.269-32.958-45.698-32.958-27.19 0-39.438 13.139-57.303 35.797-4.045 5.13-11.46 6.069-16.665 2.122l-34.699-26.31c-5.068-3.843-6.251-10.972-2.715-16.258C141.4 112.957 176.158 90 230.655 90c56.366 0 116.531 43.998 116.531 102 0 77.02-83.186 78.205-83.186 106.667z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </FontAwesomeIcon$1>
        </div>
      </CustomInput>
      <button
        type="submit"
      />
    </form>
  </Form>
</TestForm>
`;
